README.md
## Functional testing suite for SeisHub.

The goal is to treat SeisHub as a black box and just test its behaviour. This
is not meant to replace the SeisHub internal unit tests but rather augment and
complement it. Certain things are just hard/impossible to test with unit
testing, especially in a complex software package.

So far it is very simple but the plan is to enhance it quite a bit.

Currently only works with this SeisHub branch:

[https://github.com/krischer/seishub.core/tree/granular_authorization](https://github.com/krischer/seishub.core/tree/granular_authorization)

### Requirements

* seishub.core
* setuptools
* colorama

### Installation

Checkout the repository

```bash
git clone https://github.com/krischer/seishub.functional_tests
cd seishub.functional_tests
```

and then either

```bash
python setup.py install
```

or (if you prefer `pip` and have it installed)

```bash
pip install .
```

should do the trick.

### Usage

```bash
python -m seishub.functional_tests
```

By default stdout, stderr, and stdin from the SeisHub server are caught by the
test runner. To enable debugging, e.g. setting a trace or something similar,
this must not be the case. Disable it with

```bash
python -m seishub.functional_tests debug
```


CHANGELOG.md
**2012-10-17 - version 0.0.1a**
* Initial commit with the basic framework and some tests standing.


seishub_functional_tests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Functional SeisHub tests.

Rather expensive, as they always create a new SeisHub instance, launch it and
then send and receive REST request, but this tests the whole SeisHub chain.

:copyright:
    Lion Krischer (krischer@geophysik.uni-muenchen.de), 2012
:license:
    GNU Lesser General Public License, Version 3
    (http://www.gnu.org/copyleft/lesser.html)
"""
import os
import shutil
import subprocess
import sys
import time
import urllib2

from utils import print_info, print_error, TEST_DIRECTORY, BIN_DIR, BASE_URL

import test_case_vanilla_seishub
import test_case_initial_installation


def init_seishub_instance():
    """
    Creates a new SeisHub instance at TEST_DIRECTORY.
    """
    print_info("Creating new SeisHub instance...")

    if os.path.exists(TEST_DIRECTORY):
        shutil.rmtree(TEST_DIRECTORY)

    cmd = ["seishub-admin", "initenv", TEST_DIRECTORY]
    try:
        subprocess.check_output(cmd, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print_error("Error creating seishub instance. Exited with return "
            "code %s. Full output follows:" % (str(e.returncode)))
        print ""
        print e.output
        print ""
        sys.exit(1)


def launch_seishub_server():
    """
    Launches the SeisHub instance, loops until the startup sequence is finished
    and returns the process object.
    """
    print_info("Starting SeisHub Server...")

    proc = subprocess.Popen(os.path.join(BIN_DIR, "debug.sh"), shell=True,
            stdout=subprocess.PIPE, stderr=subprocess.STDOUT)

    # Ping the server for up to ten seconds. This should assure it has started.
    t = time.time()
    while (time.time() - t) < 10.0:
        try:
            urllib2.urlopen(BASE_URL).read()
            break
        except Exception, e:
            if hasattr(e, "code"):
                break
            time.sleep(0.25)

    # Give some extra time, succesful pinging means just that the webservice
    # has started.
    time.sleep(0.25)
    print_info("SeisHub server started.")
    return proc


def stop_seishub_server(proc):
    """
    Terminates the passed process object.
    """
    proc.terminate()
    proc.kill()
    proc.wait()

    # Wait for a short time and then just kill it. Quite brute force, but the
    # only reliable way I could find.
    time.sleep(0.2)
    os.system('pkill -f "python -m seishub.core.daemon"')

    print_info("SeisHub Server terminated.")


def cleanup():
    """
    Simply removes the temporarily created SeisHub instance.
    """
    print_info("Cleaning up...")

    # Try deleting it for some seconds...
    t = time.time()
    while (time.time() - t) < 10.0:
        time.sleep(0.1)
        try:
            if os.path.exists(TEST_DIRECTORY):
                shutil.rmtree(TEST_DIRECTORY)
            return
        except:
            pass
    print_error("Cleanup failed.")


if __name__ == "__main__":
    print(
        "===================================================================\n"
        "Functional tests for SeisHub.\n\n"
        "A series of tests mainly to validate the REST interface of SeisHub.\n"
        "SeisHub is largely treated as a black box and only input, output\n"
        "and HTTP response codes are tested.\n"
        "===================================================================\n"
    )
    init_seishub_instance()

    print_info("Launching test case for initial setup...")
    # Run the basic setup test suite.
    test_case_initial_installation.run()

    proc = launch_seishub_server()

    print_info("Launching test case for vanilla SeisHub installation...")
    # Run the basic setup test suite.
    test_case_vanilla_seishub.run()

    stop_seishub_server(proc)

    cleanup()


test_case_vanilla_seishub.py
import unittest
from utils import send_http_request, BASE_URL


class VanillaSeisHubTestCase(unittest.TestCase):
    """
    Simple test case that checks that the basic original setup is correct
    directly after running the initenv command.
    """
    def setUp(self):
        pass

    def test_manage_url_only_accessible_by_admin(self):
        """
        The /manage URL should only be accessible from admins.
        """
        url = BASE_URL + "/manage"
        # Admin should be able to access it.
        code, msg, req = send_http_request(url, "GET", "admin", "admin", "")
        self.assertEqual(code, 200)
        self.assertTrue(req is not None)
        # Others should be denied
        code, msg, req = send_http_request(url, "GET", "", "", "")
        self.assertEqual(code, 401)
        self.assertEqual(req, None)
        code, msg, req = send_http_request(url, "GET", "adf", "", "")
        self.assertEqual(code, 401)
        self.assertEqual(req, None)
        code, msg, req = send_http_request(url, "GET", "lo", "sd", "")
        self.assertEqual(code, 401)
        self.assertEqual(req, None)
        # Also the wrong password should be denied.
        code, msg, req = send_http_request(url, "GET", "admin", "", "")
        self.assertEqual(code, 401)
        self.assertEqual(req, None)


def run():
    suite = unittest.TestLoader().loadTestsFromTestCase(VanillaSeisHubTestCase)
    unittest.TextTestRunner(verbosity=1).run(suite)


test_case_initial_installation.py
import os
import sqlite3
import unittest

from utils import AUTH_DB, SEISHUB_DB, BIN_DIR, TEST_DIRECTORY


class InitialSetupTestCase(unittest.TestCase):
    """
    Simple test case that checks that the basic original setup is correct
    directly after running the initenv command.
    """
    def setUp(self):
        pass

    def test_files(self):
        """
        A number of files should be present in any case. Mainly the start
        scripts, databases and the config file.
        """
        files = [AUTH_DB, SEISHUB_DB,
            os.path.join(BIN_DIR, "start.sh"),
            os.path.join(BIN_DIR, "stop.sh"),
            os.path.join(BIN_DIR, "debug.sh"),
            os.path.join(TEST_DIRECTORY, "conf", "seishub.ini")]
        for file in files:
            self.assertTrue(os.path.exists(file))

    def test_check_default_user(self):
        """
        At init time, only one user should be present: admin. Check it by
        directly reading the database.

        admin is part of two groups: admin and users
        """
        conn = sqlite3.connect(AUTH_DB)
        c = conn.cursor()
        c.execute("SELECT user_name, groups from users")
        users = c.fetchall()
        self.assertEqual(len(users), 1)
        admin = users[0]
        self.assertEqual(admin[0], "admin")
        self.assertEqual(admin[1], "admin, users")

    def test_check_default_group(self):
        """
        At init time, only two groups should be present: admin and users. Check
        it by directly reading the database.
        """
        conn = sqlite3.connect(AUTH_DB)
        c = conn.cursor()
        c.execute("SELECT group_name, group_owner from groups")
        groups = c.fetchall()
        self.assertEqual(len(groups), 2)
        self.assertTrue(("admin", "admin") in groups)
        self.assertTrue(("users", "admin") in groups)


def run():
    suite = unittest.TestLoader().loadTestsFromTestCase(InitialSetupTestCase)
    unittest.TextTestRunner(verbosity=1).run(suite)


seishub/functional_tests/utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Utility methods for the SeisHub functional test suite.

:copyright:
    Lion Krischer (krischer@geophysik.uni-muenchen.de), 2012
:license:
    GNU Lesser General Public License, Version 3
    (http://www.gnu.org/copyleft/lesser.html)
"""
import colorama
import hashlib
import os
import sqlite3
import tempfile
import urllib2

__all__ = ["TEST_DIRECTORY", "BIN_DIR", "DB_DIR", "AUTH_DB", "SEISHUB_DB",
    "BASE_URL", "send_http_request", "print_info", "print_error",
    "add_user_to_seishub"]

# Make sure it works on Windows.
colorama.init()

HTTP_ACCEPTED_METHODS = ["PUT", "POST", "HEAD", "GET", "DELETE"]


def _get_installation_directory():
    tempdir = tempfile.gettempdir()
    if not tempdir:
        msg = "Could not get temporary directory."
        raise Exception(msg)
    tempdir = os.path.abspath(os.path.join(tempdir,
        "__temp_seishub_instance__"))
    return tempdir

TEST_DIRECTORY = _get_installation_directory()

BIN_DIR = os.path.join(TEST_DIRECTORY, "bin")
DB_DIR = os.path.join(TEST_DIRECTORY, "db")
AUTH_DB = os.path.join(DB_DIR, "auth.db")
SEISHUB_DB = os.path.join(DB_DIR, "seishub.db")

BASE_URL = "http://localhost:8080"


def send_http_request(url, method, user, password, data):
    """
    Simple convenience method to send (and receive from) HTTP requests.

    Return a tuple of code, message,  and request (can be None)
    """
    # Create an OpenerDirector for Basic HTTP Authentication
    password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
    password_mgr.add_password(None, url, user, password)
    auth_handler = urllib2.HTTPBasicAuthHandler(password_mgr)
    opener = urllib2.build_opener(auth_handler)
    # install globally
    urllib2.install_opener(opener)
    code, msg, request = _HTTP_request(url, method, data)
    return code, msg, request


class _RequestWithMethod(urllib2.Request):
    """
    Improved urllib2.Request Class for which the HTTP Method can be set to
    values other than only GET and POST.  See
    http://benjamin.smedbergs.us/blog/2008-10-21/\
    putting-and-deleteing-in-python-urllib2/
    """
    def __init__(self, method, *args, **kwargs):
        if method not in HTTP_ACCEPTED_METHODS:
            msg = "HTTP Method not supported. " + \
                  "Supported are: %s." % HTTP_ACCEPTED_METHODS
            raise ValueError(msg)
        urllib2.Request.__init__(self, *args, **kwargs)
        self._method = method

    def get_method(self):
        return self._method


def _HTTP_request(url, method, data):
    """
    Send a HTTP request via urllib2.

    :type url: String
    :param url: Complete URL of resource
    :type method: String
    :param method: HTTP method of request, e.g. "PUT"
    :type data: String
    :param data: XML for a send request (PUT/POST)
    """
    req = _RequestWithMethod(method=method, url=url, data=data)
    # it seems the following always ends in a urllib2.HTTPError even with
    # nice status codes...?!?
    try:
        response = urllib2.urlopen(req)
        return response.code, response.msg, response
    except urllib2.HTTPError, e:
        return e.code, e.msg, None


def print_log(color, msg):
    print color + msg + colorama.Style.RESET_ALL


def print_info(msg):
    print_log(colorama.Fore.GREEN, msg)


def print_error(msg):
    print_log(colorama.Fore.RED, msg)


def add_user_to_seishub(user_name, password, groups, real_name):
    """
    Manually adds a user to the SeisHub. This is usually only possible via the
    webinterface.

    groups should be a list of group names.
    """
    # The hash used here has to be the same as the one used in SeisHub!
    password_hash = hashlib.sha224(password).hexdigest()
    groups = ", ".join(groups)
    conn = sqlite3.connect(AUTH_DB)
    c = conn.cursor()
    c.execute("INSERT INTO users (user_name, password_hash, groups, real_name,"
        " is_active) VALUES ('%s', '%s', '%s', '%s', 1)" % (user_name,
        password_hash, groups, real_name))
    conn.commit()
    conn.close()


seishub/functional_tests/__main__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Functional SeisHub tests.

Rather expensive, as they always create a new SeisHub instance, launch it and
then send and receive REST request, but this tests the whole SeisHub chain.

:copyright:
    Lion Krischer (krischer@geophysik.uni-muenchen.de), 2012
:license:
    GNU Lesser General Public License, Version 3
    (http://www.gnu.org/copyleft/lesser.html)
"""
import os
import shutil
import subprocess
import sys
import time
import urllib2

from utils import print_info, print_error, TEST_DIRECTORY, BIN_DIR, BASE_URL, \
    add_user_to_seishub

import test_case_vanilla_seishub
import test_case_initial_installation


def init_seishub_instance(debug=False):
    """
    Creates a new SeisHub instance at TEST_DIRECTORY.

    :param debug: If debug is True, the output will not be catched.
    """
    print_info("Creating new SeisHub instance in %s..." % TEST_DIRECTORY)

    if os.path.exists(TEST_DIRECTORY):
        msg = "SeisHub temp directory already exists."
        print_error(msg)
        sys.exit(1)

    cmd = ["seishub-admin", "initenv", TEST_DIRECTORY]
    try:
        if debug is True:
            subprocess.call(cmd)
        else:
            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print_error("Error creating seishub instance. Exited with return "
            "code %s. Full output follows:" % (str(e.returncode)))
        print ""
        print e.output
        print ""
        sys.exit(1)


def launch_seishub_server(debug=False):
    """
    Launches the SeisHub instance, loops until the startup sequence is finished
    and returns the process object.

    :param debug: If debug is True, the output will not be catched.
    """
    print_info("Starting SeisHub Server...")

    if debug is True:
        proc = subprocess.Popen(os.path.join(BIN_DIR, "debug.sh"), shell=True)
    else:
        proc = subprocess.Popen(os.path.join(BIN_DIR, "debug.sh"), shell=True,
                stdout=subprocess.PIPE, stderr=subprocess.STDOUT)

    # Ping the server for up to ten seconds. This should assure it has started.
    t = time.time()
    while (time.time() - t) < 10.0:
        try:
            urllib2.urlopen(BASE_URL).read()
            break
        except Exception, e:
            if hasattr(e, "code"):
                break
            time.sleep(0.25)

    # Give some extra time, succesful pinging means just that the webservice
    # has started.
    time.sleep(0.25)
    print_info("SeisHub server started.")
    return proc


def stop_seishub_server(proc):
    """
    Terminates the passed process object.
    """
    proc.terminate()
    proc.kill()
    proc.wait()

    # Wait for a short time and then just kill it. Quite brute force, but the
    # only reliable way I could find.
    time.sleep(0.2)
    os.system('pkill -f "python -m seishub.core.daemon"')

    print_info("SeisHub Server terminated.")


def cleanup():
    """
    Simply removes the temporarily created SeisHub instance.
    """
    print_info("Cleaning up...")

    # Try deleting it for some seconds...
    t = time.time()
    while (time.time() - t) < 10.0:
        time.sleep(0.1)
        try:
            if os.path.exists(TEST_DIRECTORY):
                shutil.rmtree(TEST_DIRECTORY)
            return
        except:
            pass
    print_error("Cleanup failed.")


if __name__ == "__main__":
    print(
        "===================================================================\n"
        "Functional tests for SeisHub.\n\n"
        "A series of tests mainly to validate the REST interface of SeisHub.\n"
        "SeisHub is largely treated as a black box and only input, output\n"
        "and HTTP response codes are tested.\n"
        "===================================================================\n"
    )

    if "debug" in sys.argv:
        debug = True
    else:
        debug = False

    # Create a new SeisHub instance.
    init_seishub_instance(debug=debug)

    print_info("Launching test case for the initial instance setup...")
    test_case_initial_installation.run()

    # Add two more users. A normal one, and one with admin rights.
    add_user_to_seishub("admin_2", "admin_pw", ["users", "admin"], "New Guy")
    add_user_to_seishub("user", "user_pw", ["users"], "Some user")

    # Launch the server.
    proc = launch_seishub_server(debug=debug)

    print_info("Launching test case for the vanilla SeisHub server...")
    test_case_vanilla_seishub.run()

    # Stop/kill the server.
    stop_seishub_server(proc)

    cleanup()
